src/tests/nativeaot/SmokeTests/TrimmingBehaviors/DeadCodeElimination.cs

```diff
@@ -330,21 +331,80 @@ public static void Run()
         }
     }
 
+    class TestBranchesInGenericCodeRemoval
+    {
+        struct Unused { public byte Val; }
+        struct Used { public byte Val; }
+
+        [MethodImpl(MethodImplOptions.NoInlining)]
+        static T Cast<T>(byte o)
+        {
+            if (typeof(T) == typeof(Unused))
+            {
+                Unused result = new Unused { Val = o };
+                return (T)(object)result;
+            }
+            else if (typeof(T) == typeof(Used))
+            {
+                Used result = new Used { Val = o };
+                return (T)(object)result;
+            }
+            return default;
+        }
+
+        public static void Run()
+        {
+            Console.WriteLine("Testing dead branches guarded by typeof in generic code removal");
+
+            Cast<Used>(12);
+
+            // We only expect to be able to get rid of it when optimizing
+#if !DEBUG
+            ThrowIfPresentWithTypeHandle(typeof(TestBranchesInGenericCodeRemoval), nameof(Unused));
+#endif
+            ThrowIfNotPresent(typeof(TestBranchesInGenericCodeRemoval), nameof(Used));
+        }
+    }
+
     [UnconditionalSuppressMessage("ReflectionAnalysis", "IL2070:UnrecognizedReflectionPattern",
         Justification = "That's the point")]
-    private static bool IsTypePresent(Type testType, string typeName) => testType.GetNestedType(typeName, BindingFlags.NonPublic | BindingFlags.Public) != null;
+    private static Type GetTypeSecretly(Type testType, string typeName) => testType.GetNestedType(typeName, BindingFlags.NonPublic | BindingFlags.Public);
 
     private static void ThrowIfPresent(Type testType, string typeName)
     {
-        if (IsTypePresent(testType, typeName))
+        if (GetTypeSecretly(testType, typeName) != null)
         {
             throw new Exception(typeName);
         }
     }
 
+    private static void ThrowIfPresentWithTypeHandle(Type testType, string typeName)
+    {
+        Type t = GetTypeSecretly(testType, typeName);
+        if (t == null)
+        {
+            throw new Exception("Not found " + typeName);
+        }
+
+        bool thrown = false;
+        try
+        {
+            _ = t.TypeHandle;
```

Comment:
We still keep the metadata around because metadata got figured out during scanning, but we can check for the presence of the `MethodTable` by looking at the `TypeHandle` property.

The `Type` is damaged like this because we obtained it through unsafe reflection code. Legit code should not be able to see damaged types like this.